sycl/plugins/unified_runtime/ur/adapters/hip/enqueue.cpp

```diff
@@ -896,21 +1016,32 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueMemImageRead(
     ur_rect_offset_t origin, ur_rect_region_t region, size_t, size_t,
     void *pDst, uint32_t numEventsInWaitList,
     const ur_event_handle_t *phEventWaitList, ur_event_handle_t *phEvent) {
-  UR_ASSERT(hImage->MemType == ur_mem_handle_t_::Type::Surface,
-            UR_RESULT_ERROR_INVALID_MEM_OBJECT);
+  UR_ASSERT(hImage->isImage(), UR_RESULT_ERROR_INVALID_MEM_OBJECT);
+  ur_image_ *Image = ur_cast<ur_image_ *>(hImage);
 
   ur_result_t Result = UR_RESULT_SUCCESS;
 
+  ur_lock MemoryMigrationLock(hImage->MemoryMigrationMutex);
+
+  // Note that this entry point may be called on a specific queue that may not
+  // be the last queue to write to the MemBuffer
+  auto DeviceToCopyFrom = Image->LastEventWritingToMemObj == nullptr
+                              ? hQueue->getDevice()
+                              : Image->LastEventWritingToMemObj->getDevice();
+
   try {
-    ScopedContext Active(hQueue->getDevice());
-    hipStream_t HIPStream = hQueue->getNextTransferStream();
+    ScopedDevice Active(DeviceToCopyFrom);
+    // Use the default stream if copying from another device
+    hipStream_t HIPStream = DeviceToCopyFrom == hQueue->getDevice()
+                                ? hQueue->getNextTransferStream()
+                                : hipStream_t{0};
```

Comment:
How does this work for `hipMemcpyDToD` then? does the stream have to be for the origin device and not for the destination device?

Reply:
EnqueueMemBufferRead is only to read a buffer to the host, so a DtoD is never needed, only a DtoH from the most up to date device

Reply:
So DtoH requires a stream from that particular device?